Initial commit

notalfredo opened the pull request:
- **Please check if the PR fulfills these requirements**
   - [ ] The commit message follows our guidelines
   - [ ] Tests for the changes have been added (for bug fixes / features)
   - [ ] Docs have been added / updated (for bug fixes / features)
 - **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
 - **What is the current behavior?** (You can also link to an open issue here)
 - **What is the new behavior (if this is a feature change)?**
 - **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
 - **Other information**:

notalfredo committed:
Initial commit

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,560 @@
+                &y.to_string())
+            },
+            _ => {
+                panic!("Unrecoginzed Opperation")
```
Typo, should be "Operation"

neoncitylights replied:
There should be a match arm for `StringDiffOpKind::Transpose`, and since that's not currently not supported you can panic with a message similar to "applyDiff() does not currently support the transpose operation yet." ("Opperation" is also a typo). 

Since the `match` construct can do [exhaustive matching](https://doc.rust-lang.org/book/ch06-02-match.html#matches-are-exhaustive), you can get rid of the default match arm (`_`) after adding one for `Transpose`.

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,560 @@
+}
+
+#[derive(Debug)]
+pub struct StringDiffOp {
+    pub kind: StringDiffOpKind,
+    pub index: isize,
+}
```
There should be a constructor for this type, which should look like:
```rs
impl StringDiffOp {
   pub fn new(kind: StringDiffOpKind, index: isize) {
       Self { kind, index }
   }
}
```

And any instance of creating a new `StringDiffOp` should use this, via `StringDiffOp::new()`

neoncitylights replied:
Could you also derive the `PartialEq` and `Eq` traits for StringDiffOp? This allows overloading the `==` and `!=` operators and comparing them against each other.

For the `index` property: since `isize` is signed, it doesn't really make sense because string indices are `usize` (unsigned, zero or positive), where as signed would imply possibly negative indices. So, this should be the `usize` type.

----

There should be a constructor for the `StringDiffOp` type, which would look like:
```rs
impl StringDiffOp {
   // a generic constructor
   pub fn new(kind: StringDiffOpKind, index: usize) -> Self {
       Self {
           kind,
           index
       }
   }
   
   // utility method to create a delete-specific StringDiffOp
   pub fn new_delete(c: char, index: usize) -> Self {
       Self::new(StringDiffOpKind::Delete(c), index)
   }
}
```

And then callers can create a new instance of `StringDiffOp` via `StringDiffOp::new()`, and pass in the appropriate arguments. Same with `StringDiffOp::new_delete()`, `StringDiffOp::new_insert()`, etc.

The utility constructors also make it easier since otherwise would be a bit verbose+repetitive to keep doing the same thing. Compare:

```rust
// original
StringDiffOp {
    kind: StringDiffOpKind::Delete('c')
    index: 1usize,
}

// second (but StringDiffOpKind::new() can still be public)
StringDiffOpKind::new(StringDiffOp::Delete('c'), 1usize)

// now
StringDiffOpKind::new_delete('c', 1usize)
```

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,560 @@
+#[derive(Debug)]
+pub struct StringDiffOp {
+    pub kind: StringDiffOpKind,
+    pub index: isize,
```
Is there a specific reason this is an `isize` (signed)? This would make more sense as a `usize` (unsigned) since you can't negatively index a string

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,560 @@
+}
+
+pub struct HammingDistance {}
+impl HammingDistance{}
```
Since this is an empty `impl`, we can remove this

neoncitylights replied:
Since this `impl` block is empty, we can safely remove this

neoncitylights replied:
Since this is an empty `impl` block, this can be safely removed

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,560 @@
+}
+
+pub trait StringDiffAlgorithm {
+    //fn diff<'a>(s1: &'a str, s2: &'a str) -> Vec<StringDiffOp>;
```
This commented out line can be safely removed

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,560 @@
+#[cfg(test)]
+mod dcode_tests{
+
+    fn eq_with_nan_eq(a: &StringDiffOp, b: &StringDiffOp) -> bool {
```
By deriving the equality-related traits for `StringDiffOp` (mentioned above), you could just do `b == a` and that'd work, and you can take out this function since it won't be necessary anymore

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,560 @@
+        assert_eq!(vec_compare(&test_vec_4, &test_struct.diff("0000", "1111")) , true );
+        assert_eq!(vec_compare(&test_vec_5, &test_struct.diff("2173896", "2233796")) , true );
+
+        //assert_eq!( (test_vec_5[0].kind).0 ,super::StringDiffOpKind::Substitute('7', '3'));
```
Why was this commented out?

notalfredo replied:
Ignore that junk ill delete it (-｡-;

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,560 @@
+}
+
+pub (crate) fn remove(start: usize, stop: usize, s: &str) -> String {
+    let mut rslt = "".to_string();
```
- At the beginning of the function, there should be a preliminary check to make sure stop is greater than start. This error checking will cause a panic for the caller if they don't pass in the correct arguments.
   ```rust
   assert!(stop > start);
   ```
 - The variable name should be `result` since it's more readable
 - We can a create a new empty `String` type immediately via `String::new()`. This is nice since it also doesn't require changing types; `"".to_string()` creates a `&str` and then converts it to `String`

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,560 @@
+
+    
+    fn distance<'a>(&self, s1: &'a str, s2: &'a str) -> usize{
+        let mut edit_distance = 0;
```
This statement initializing the `edit_distance` variable should go after the error checking to avoid allocating memory whenever the caller hits that error. It should also specify the type via `0usize`

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,560 @@
+pub struct LevenshteinDistance {}
+impl LevenshteinDistance{
+    pub(crate) fn min_dist(x: usize, y: usize, z: usize) -> usize {
+        if x <= y && x <= z{
+            return x;
+        }
+        else if y <= x && y <= z {
+            return y;
+        }
+        else{
+            return z;
+        }
```
If I'm understanding this correctly, this tries to find the smallest number between three numbers, right? This can be simplified with two calls to `std::cmp::min()` ([docs](https://doc.rust-lang.org/std/cmp/fn.min.html)), like:

```rust
use std::cmp::min;

pub(crate) fn min_dist(x: usize, y: usize, z: usize) -> usize {
    min(x, min(y, z))
}
```

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,560 @@
+
+        let mut test_vec: Vec<StringDiffOp> = Vec::new();
+        test_vec.push(
+                super::StringDiffOp{
```
These calls can be made more succinct after creating the utility constructors, e.g this could be `StringDiffOp::new_substitute()`

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,560 @@
+                                    left_string: &str,
+                                    top_string: &str) -> Vec<StringDiffOp> {
+                        
+        let mut my_opp_vector : Vec<StringDiffOp> = Vec::new();
```
I think this would make more sense to be called `diff_ops` instead (operations also only has 1 p)

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,560 @@
+            match my_opp[left_str_len][top_str_len] {
+                //insertion
+                '^' => {
+                    let insertionOpp : StringDiffOp = StringDiffOp{
```
this should be `insertion_op` (variables are traditionally `camel_case`), also only 1 p, same for the instance below

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,560 @@
+                    }
+ 
+                    if left_string.chars().nth(left_str_len-1).unwrap() != top_string.chars().nth(top_str_len-1).unwrap(){
+                        let insertionOpp : StringDiffOp = StringDiffOp{
```
Copy/paste error? :P This should be `substitute_op`

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,560 @@
+            }
+        }
+
+        //LevenshteinDistance::print_vector(&dist_vector);
+        //LevenshteinDistance::print_vector(&dir_vector);
+        /*
+        for row in dist_vector.iter(){
+            println!("{:?}", row)
+        }
+        for row in dir_vector.iter(){
+            println!("{:?}", row)
+        }
+        */
+
+        //println!("Made it here");
+        let test = LevenshteinDistance::get_operations(&dir_vector, s1, s2);
+        println!("{:?}", test);
+        //return dist_vector[first_string_len][second_string_len];
+        return test;
```
There's some debug stuff leftover that should be taken out, and then immediately return `LevenshteinDistance::get_operations(&dir_vector, s1, s2)` without creating a new variable

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,560 @@
+        }
+    }
+
+    pub(crate) fn min_dist_with_dir(x: usize, y: usize, z: usize) -> (usize,char){
```
Could you explain a little bit about what this function does? I'm assuming `dir` means direction here? I can see that down below each one represents a kind of diff op, but each character seems arbitrarily chosen and can be a little hard to follow until you look down below. This is also a good opportunity to write a docblock for the function with `///`

Maybe with some clarification we can make this a bit more readable! 👍

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,560 @@
+                    if prev_char == '^'{
+                        my_opp_vector.reverse();
+                    }
+                    let insertionOpp : StringDiffOp = StringDiffOp{
```
`delete_op`

neoncitylights reviewed (changes requested):
Thank you so much for writing this! :D Overall, I really like the direction of this code. I wrote some initial comments for the biggest things that stick out to me.

Can you run `rustfmt` and `clippy` as well? `Rustfmt` is the code-formatting tool, and `clippy` acts as the linter (think of like ESLint and Prettifier)

notalfredo committed:
Second commit

Finished mostly all issues | Fixed some stuff with Rustfmt and clippy still need to do more tho

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,458 @@
+	/// if we chooze x(The value above the cell) as the min value its a insertion operation (symbolized by '^')
+	/// if we choose y(The value left of the cell) as the min value its a deletion operation(symbolized by '<')
+	/// if we choose z(The value diagnal of the cell) as the min value its a substitution operation( sybmolized by '\' )
+	/// else we get an x (should never happen since either(x,y,z should be the smalles))
```
>else we get an x (should never happen since either(x,y,z should be the **smalles**))

typo, should be "smallest". Also, since that condition should never occur, then you can replace the line `(0, 'x')` with the `unreachable!()` macro ([docs](https://doc.rust-lang.org/std/macro.unreachable.html)).

notalfredo committed:
Third commit

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,455 @@
+			}
+			if top_str_len == 0 && left_str_len == 0 {
+				break;
+			}
```
This block of code can be a little bit more succinct, since there's a few conditions checked twice (checking the length of both strings).
```rs
if top_str_len == 0 && left_str_len == 0 {
	break;
}

if prev_char == '^' {
	Self::reverse_vec_and_indexes(&mut diff_ops, top_str_len);
	break;
}
```

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,455 @@
+	result
+}
+
+pub fn aplly_diff(s: &str, diffs: Vec<StringDiffOp>) -> String {
```
Sorry I didn't catch this, this is a typo. It should be "apply" (not aplly)

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,455 @@
+				new_string.replace_range((i.index)..(i.index + 1), &_y.to_string())
+			}
+			StringDiffOpKind::Transpose(_x, _y) => {
+				panic!("aplly_diff does not currently support the transpose operation yet")
```
`aplly_diff` -> `apply_diff()`

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,455 @@
+		cmp::min(x, cmp::min(y, z))
+	}
+	/// At a given (x,y) we must choose the minimum value between a cells
+	/// Top, Left, and Diagnal value. Depending on which cell is chosen between
```
`Diagnal` -> `Diagonal`

neoncitylights reviewed (changes requested):
This will be the last review for this set of code! Just minor changes, and I'll go ahead and merge this after. I'll open up some issues so we can follow-up! Thanks a bunch.

neoncitylights commented on src/lib.rs:
```diff
@@ -1,6 +1,455 @@
+}
+
+impl StringDiffOp {
+	// a generic constructor
```
Sorry, I meant this comment as an explanation to you in the code review (to help explain the difference between the constructors); we can delete this comment

notalfredo committed:
Final Commit - From version_pancake
